ninchatsdk/src/main/java/com/ninchat/sdk/ninchatquestionnaire/ninchatquestionnairelist/presenter/NinchatQuestionnaireListPresenter.kt

```diff
@@ -0,0 +1,143 @@
+            value?.let {
+                putOpt("result", value)
+            } ?: remove("result")
+            putOpt("hasError", hasError)
+        }
+    }
+
+    override fun onUpdate(value: Boolean, sublistPosition: Int, hasError: Boolean, position: Int) {
+        model.answerList.getOrNull(position)?.apply {
+            optJSONArray("options")?.optJSONObject(sublistPosition)?.apply {
+                putOpt("result", value)
+                putOpt("hasError", hasError)
+            }
+        }
+    }
+
+    override fun onUpdate(value: String?, hasError: Boolean, position: Int) {
+        model.answerList.getOrNull(position)?.apply {
+            value?.let {
+                putOpt("result", value)
+            } ?: remove("result")
+            putOpt("hasError", hasError)
+        }
+    }
+
+    // A sentinel value to make sure we don't fall into infinite loop
+    protected fun getNextElement(currentIndex: Int, sentinel: Int): String? {
+        if (sentinel < 0) return null
+        val currentElement = model.questionnaireList.getOrNull(currentIndex)
+        return when {
+            NinchatQuestionnaireType.isLogic(currentElement) -> {
+                // is this logic a match for all any existing answer ?
+                val matches = NinchatQuestionnaireJsonUtil.matchAnswerList(logicElement = currentElement, answerList = model.answerList, preAnswerList = model.preAnswers)
+                if (matches) {
+                    model.updateTagsAndQueueId(logicElement = currentElement?.optJSONObject("logic"))
+                    currentElement?.optJSONObject("logic")?.optString("target")
+                } else
+                    this.getNextElement(currentIndex = currentIndex + 1, sentinel = sentinel - 1)
+            }
+            NinchatQuestionnaireType.isElement(currentElement) -> {
+                currentElement?.optString("name")
+            }
+            else -> {
+                // null(end of element), or not an element or logic
+                null
+            }
+        }
+    }
+
+    protected fun loadNextByElementName(elementName: String?): Int {
+        val index = model.getIndex(elementName = elementName)
+        val nextElement = model.questionnaireList.getOrNull(index)
+        return model.addElement(jsonObject = nextElement)
+    }
+
+
+    fun getAnswerList(): List<JSONObject> {
+        val uniquePreAnswers = model.preAnswers.filter { currentAnswer ->
+            val name = currentAnswer.optString("name")
+            model.answerList.any {
+                name == it.optString("name")
+            }.not()
+        }
+        return model.answerList.plus(uniquePreAnswers)
+    }
+
+    override fun onCompleteLoading(target: String?, thankYouText: String?, loaded: Boolean, position: Int) {
+        model.answerList.getOrNull(position)?.apply {
+            putOpt("loaded", true)
+        }
+    }
+
+    open fun init() {
+        TODO("implement me")
```

Comment:
What about these?

Reply:
They are intentional. :D 
They are methods of an abstract class. In kotlin TODO is `NotImplementedError`.https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-t-o-d-o.html

It is excepted that the class that implementing this abstract class, must implement those methods. There is no particular reason why I follow this pattern ( just a better note that these methods must be implemented by child class ). It seems I inherited this practice from some other projects.